Support delayed messages on the ToastMessengerService

acelaya opened the pull request:
This PR introduces support to schedule delayed messages via `ToastMessagesService`, which are displayed only if the frame where the service is running is focused, or otherwise delayed until it is focused.

This can be useful for when we want to toast hidden messages which announce the availability of a keyboard shortcut that would not work unless the frame is focused.

I have considered three approaches to do this:

* Extend the `ToastMessagesService` to support a new `delayed: boolean` message option, and wrap the logic to handle the queue and the window events internally.
  * Good because implicitly exposes the new behavior anywhere the service is already being used.
  * Good because with one dependency we can "display" messages right away or delay them.
  * Bad because it makes `ToastMessagesService` more complex, and it is possibly handling more than one responsibility.
* Create a new service wrapping `ToastMessagesService`, which implicitly always delays messages.
  * Good because it allows to keep the existing `ToastMessagesService` untouched.
  * Good because the new service will just have the new logic, and rely on `ToastMessagesService` to do the actual notification publishing.
  * Bad because any consumer which needs to publish both delayed and not delayed messages now needs two dependencies.
* Create a new service wrapping `ToastMessagesService`, which allows to toast regular messages, or delayed ones (via new option or a new set of public methods).
  * Good because it allows to keep the existing `ToastMessagesService` untouched.
  * Good because with one dependency we can "display" messages right away or delay them.
  * Bad because wherever we need delayed messages, we will need to replace the dependencies.
  * Good because it should be possible to replace the existing `ToastMessagesService` with this new service, wherever it is used.

robertknight reviewed (commented):
I had a look and left some notes. The general approach looks good.

acelaya commented on src/sidebar/services/toast-messenger.ts:
```diff
@@ -30,9 +46,25 @@ type MessageOptions = {
+   * This just holds a set of delayed messages that need to be published as soon
+   * as the window gets focused
+   */
+  private _delayedMessagesQueue: Set<MessageData>;
```
Do you think so? 🤔. To me it sounds more natural in plural, as it is a "queue containing delayed messages".

acelaya replied:
Ok, never mind. I think I actually get your point here.

The plural would make more sense if it was called just `delayedMessages`, but by naming it queue, then `delayedMessageQueue` is more logical, like in `messageRepository`, `messageStore`, etc.

robertknight replied:
> Do you think so? 🤔

Yes. As a point of comparison, grep for similar variables in the code:

```sh
ag '[a-z](Set|List|Queue)\b' src/
```

This turns up `BucketSet`, `ThreadList` etc.

I'm not certain of the proper explanation, but I think it would be something like this: "message" is referring to the kind of object in the "queue", rather than the number.

acelaya commented on src/sidebar/services/toast-messenger.ts:
```diff
@@ -60,19 +92,29 @@ export class ToastMessengerService {
     }
 
+    // Delayed messages should just be enqueued when the document is not focused
+    if (delayed && !this._window.document.hasFocus()) {
```
What do you mean? Are you asking if messages delayed while the tab is hidden will be properly pushed when the tab becomes active, or if we should be delaying messages at all while the tab is hidden?

robertknight replied:
I was just wondering aloud how tab visibility would affect this. For the initial use case of reporting when a highlight is created it doesn't actually matter since the tab will (almost?) definitely be visible when the highlight is created.

acelaya commented on src/sidebar/services/test/toast-messenger-test.js:
```diff
@@ -13,9 +15,17 @@ describe('ToastMessengerService', () => {
       updateToastMessage: sinon.stub(),
     };
+    fakeWindow = {
+      addEventListener: sinon.stub().callsFake((_, listenerCallback) => {
```
I like the idea of using an actual EventTarget!

acelaya commented on src/sidebar/services/toast-messenger.ts:
```diff
@@ -20,6 +20,22 @@ type MessageOptions = {
+  delayed?: boolean;
+};
+
+type MessageType = 'error' | 'success' | 'notice';
```
I'm going to take the opportunity to export the existing `MessageOptions` type, as I forgot to do it when I migrated this to TypeScript

acelaya commented:
@robertknight I have addressed all your comments, except for this one https://github.com/hypothesis/client/pull/5323/files#r1136876689, as I'm not completely sure if I'm understanding it correctly.

acelaya commented:
> I was just wondering aloud how tab visibility would affect this. For the initial use case of reporting when a highlight is created it doesn't actually matter since the tab will (almost?) definitely be visible when the highlight is created.

Oh, I see. I will try this, as the first use case is not actually about highlights, but new annotations available, and that can happen at any moment without user interaction.

acelaya commented:
> > I was just wondering aloud how tab visibility would affect this. For the initial use case of reporting when a highlight is created it doesn't actually matter since the tab will (almost?) definitely be visible when the highlight is created.
> 
> Oh, I see. I will try this, as the first use case is not actually about highlights, but new annotations available, and that can happen at any moment without user interaction.

Si I have just tried, and delayed messages get indeed enqueued while the tab is not active, and once you are back and the frame is focused, you get notified. This does not happen with regular notifications.

Should we cover this case and just "discard" delayed messages if the "tab is not active"?
